app/master/slave.py

```diff
@@ -244,6 +250,87 @@ def update_last_heartbeat_time(self):
     def get_last_heartbeat_time(self) -> datetime:
         return self._last_heartbeat_time
 
+    def _update_slave_registry(self):
+        """
+        Remove shutdown-ed slave from SlaveRegistry.
+        """
+        self._logger.info('Removing slave (url={}; id={}) from Slave Registry.'.format(self.url, self.id))
+        SlaveRegistry.singleton().remove_slave(slave_url=self.url)
+
+
+class SlaveRegistry(Singleton):
+    """
+    SlaveRegistry class is a singleton class which stores and maintains list of connected slaves.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.all_slaves_by_url = {}
+        self.all_slaves_by_id = {}
+        self._slave_dict_lock = Lock()
+
+    def add_slave(self, slave):
+        """
+        Add slave in SlaveRegistry.
+
+        :type slave: Slave
+        """
+        with self._slave_dict_lock:
+            self.all_slaves_by_url[slave.url] = slave
+            self.all_slaves_by_id[slave.id] = slave
+
+    def remove_slave(self, slave=None, slave_url=None):
+        """
+        Remove slave from the both url and id dictionary.
+
+        :type slave: Slave
+        :type slave_url: str
+        """
+        if (slave is None) == (slave_url is None):
+            raise ValueError('Only one of slave or slave_url should be specified to remove_slave().')
+
+        with self._slave_dict_lock:
+            try:
+                if slave is None:
+                    slave = self.get_slave(slave_url=slave_url)
+                del self.all_slaves_by_url[slave.url]
+                del self.all_slaves_by_id[slave.id]
+            except (ItemNotFoundError, KeyError):
+                # Ignore if slave does not exists in SlaveRegistry
+                pass
+
+    def get_slave(self, slave_id=None, slave_url=None):
+        """
+        Look for a slave in the registry and if not found raise "ItemNotFoundError" exception.
+
+        :param slave_id: The id of the slave to return
+        :type slave_id: int
+        :param slave_url: The url of the slave to return
+        :type slave_url: str
+        :return: The instance of the slave
+        :rtype: Slave | None
+        """
```

Comment:
The rtype of this function is just `Slave`, not `Slave | None`

I think it _should_ be `Slave | None` as I argued above, but currently it's not.

Reply:
What if its either Slave or ItemNotFound Exception? I am doing `Union[Slave, ItemNotFoundError]`. Is that correct?